File: 0402-remove-k-digits/0402-remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        l = len(num)
        if l == k : return '0'
        stack = []
        for x in num:
            while stack and k and  stack[-1]> x:
                stack.pop()
                k-=1
            if stack or x != '0':
                stack.append(x)
        if k :
            stack = stack[:-k]
        return ''.join(stack) or '0'
        

File: 0402-remove-k-digits/0402-remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
    vector<char> st;  
    for (char c : num) {
        while (!st.empty() && k > 0 && st.back() > c) {
            st.pop_back();
            k--;
        }
        st.push_back(c);
    }

    while (k > 0) {   
        st.pop_back();
        k--;
    }

    string ans = "";
    for (char c : st) ans.push_back(c);

    int i = 0;
    while (i < ans.size() && ans[i] == '0') i++;

    ans = ans.substr(i);
    return ans == "" ? "0" : ans;
}

};